03week/towersOfHanoi.js

```diff
@@ -13,33 +13,129 @@ let stacks = {
+// returns true/false
+// TRUE: if input is: a, b, c
+// FALSE: otherwise
+function validateInput(startStack, endStack) {
+  const start = sanitizeInput(startStack);
+  const end = sanitizeInput(endStack);
+
+  return ((start === 'a' || start === 'b' || start === 'c') && (end === 'a' || end === 'b' || end === 'c'));
+}
+
+// prints the current stacks object
 function printStacks() {
   console.log("a: " + stacks.a);
   console.log("b: " + stacks.b);
   console.log("c: " + stacks.c);
 }
 
-function movePiece() {
-  // Your code here
-
+// moves a piece from one stack to the other
+// will only move piece if the move isLegal
+// does not return anything
+function movePiece(startStack, endStack) {
+  // sanitize user input
+  const start = sanitizeInput(startStack);
+  const end = sanitizeInput(endStack);
+
+  // if is a legal move, then move piece
+  if ( isLegal(start, end) ) {
+    const piece = stacks[start].pop();
+    stacks[end].push(piece);
+  }
 }
 
-function isLegal() {
-  // Your code here
-
+// determines whether a move is legal or not
+// returns true/false
+// TRUE: if stack is empty OR if current piece on endstack is greater size than piece being moved
+// FALSE: if piece being moved is greater than piece on endstack
+function isLegal(startStack, endStack) {
+  // Sanitize user input
+  const start = sanitizeInput(startStack);
+  const end = sanitizeInput(endStack);
+
+  // get the lengths of the start and end arrays
+  const startLen = stacks[start].length;
+  const endLen = stacks[end].length;
+
+  // if stack is empty, then good to move piece
+  if (endLen === 0) {
+    return true;
+  } else {
+    // if current piece on endstack is greater size than piece being moved
+    return (stacks[end][endLen - 1] > stacks[start][startLen - 1]);
+  }
 }
 
+// checks for win
+// win = pieces placed largest -> smallest on either b or c stack
+// returns true/false
 function checkForWin() {
-  // Your code here
+  return ((stacks.c[0] === 4 && stacks.c[1] === 3 && stacks.c[2] === 2 && stacks.c[3] === 1) || (stacks.b[0] === 4 && stacks.b[1] === 3 && stacks.b[2] === 2 && stacks.b[3] === 1));
+}
 
+// win banner to display once a user wins
+function winBanner() {
+  console.log("\n========================");
+  console.log("WIN!!!");
+  console.log("========================\n");
 }
 
-function towersOfHanoi(startStack, endStack) {
-  // Your code here
+// resets the global vars (and therefor the game)
+function resetGame() {
+  stacks = {
+    a: [4, 3, 2, 1],
+    b: [],
+    c: []
+  };
+}
+
+/*
+  WHITEBOARD
 
+  [x] sanitize input
+  [x] validate move - isLegal()
+  [x] IF: legal move, then movePiece()
+  [x] ELSE: console log illegal move
+
+
+ */
+
+// main function
+function towersOfHanoi(startStack, endStack) {
+  const start = sanitizeInput(startStack);
+  const end = sanitizeInput(endStack);
+
+  // if user input is valid
+  if ( validateInput(start, end) ) {
+    // if stack move is a legal move
+    if ( isLegal(start, end) ) {
```

Comment:
here, you're checking if it's legal correctly, so the check for legality in move piece is once again redundant